### PR TITLE
ApiPort for .NET Core (including x-plat console host)

### LINKED ISSUE #117: Switch ApiPort to use DNX
Currently ApiPort targets .NET Framework 4.6, but we could switch it to use DNX on .NET Core so it doesn't have the 4.6 requirement.

### PR DESCRIPTION
**Pull request by mjrousos**
This adds a CoreCLR ApiPort build. The changes to ApiPort itself are fairly straightforward, though the magic of a csproj that builds as a .NET Core exe was more complicated.

A few things to know about this PR -
* This upgrades System.Reflection.Metadata (because the version we were using previously did not support .NET Core). I had to use a very new alpha version to get one that both targeted .NET Core and had all the features we need. 
* With these changes, it's necessary to disable NuGet restore in VS (as per [these](http://aka.ms/port2core) instructions) or there will be some false errors (the error list shows issues, but the build completes successfully). NuGet restore can be left on if you don't mind noise, but otherwise restoring from the command line works cleanly.
* Because the CoreCLR ApiPort host is OS- and architecture- specific, there are a lot more configurations in the sln file now. For everything except ApiPort.Core, the projects build the same for all of them. That project needs to target each individually, though, in order to get the correct hosting bits.
* This has been tested on Linux (Ubuntu 14.04) but is not yet tested on Mac (my MacBook is in the mail on its way to me currently). There may be a few Mac-specific tweaks necessary.
* Because System.Configuration isn't available for .NET Core, Unity configuration via config file is only enabled for Desktop ApiPort. We should look into run-time configuration through alternate means for the future. I think it's ok to merge this as is in the short-term, though. It just means that offline-mode is a to-be-implemented feature for ApiPort.Core at this point.

Fixes #117

@twsouthwick @conniey

### CODE REVIEW

**Comment by twsouthwick** [comment]
Would it be better to have an ApiPort.Core solution that's separate and pulls in the nuget packages it needs?

**Comment by twsouthwick** [comment]
As far as Unity stuff is concerned, we could switch it to use Autofac, which has a nice, modern configuration system available

File: src/ApiPort/ApiPort.csproj [hunk 1]
```diff
@@ -13,6 +13,7 @@
     <TargetFrameworkVersion>v4.5</TargetFrameworkVersion>
     <FileAlignment>512</FileAlignment>
     <AutoGenerateBindingRedirects>true</AutoGenerateBindingRedirects>
+    <DefineConstants>DESKTOP_CONFIGURATION;RICH_CONSOLE</DefineConstants>
```
Comment (twsouthwick) [inline src/ApiPort/ApiPort.csproj]: The rest of the solution uses the convention FEATURE_* for define constants (such as FEATURE_DESKTOP_CONFIGURATION and FEATURE_RICH_CONSOLE). Can you switch these to follow that?

File: src/ApiPort/ApiPort.project.json [hunk 2]
```diff
@@ -8,8 +8,8 @@
     "Microsoft.Framework.Configuration.CommandLine": "1.0.0-beta5",
     "Microsoft.Framework.OptionsModel": "1.0.0-beta5",
     "Newtonsoft.Json": "6.0.8",
-    "System.Collections.Immutable": "1.1.37",
-    "System.Reflection.Metadata": "1.1.0-alpha-00008",
+    "System.Collections.Immutable": "1.1.38-*",
+    "System.Reflection.Metadata": "1.2.0-alpha-00017",
```
Comment (twsouthwick) [inline src/ApiPort/ApiPort.project.json]: :) @conniey  was looking into updating this but there was a bug in the metadata reader she found that was fixed in this release.

File: src/ApiPort/CommandLineOptions.cs [hunk 3]
```diff
@@ -70,9 +70,17 @@ private static ICommandLineOptions ShowHelp(string suppliedCommand = null, bool
                 // TODO: Get invalid parameter (Microsoft.Framework.Configuration currently does not surface this)
                 Program.WriteColorLine($"Invalid parameter passed to {suppliedCommand}", ConsoleColor.Red);
             }
-
-            var codebase = new Uri(typeof(CommandLineOptions).GetTypeInfo().Assembly.CodeBase);
-            var path = Path.GetFileName(codebase.AbsolutePath);
+            
+            var location = typeof(CommandLineOptions).GetTypeInfo().Assembly.Location;
+            var path =
+#if NETCORE
```
Comment (twsouthwick) [inline src/ApiPort/CommandLineOptions.cs]: Can you use FEATURE_ASSEMBLY_LOCATION? I think this is already defined (maybe in a different project, though)

File: src/ApiPort/ConsoleProgressReporter.cs [hunk 4]
```diff
@@ -77,6 +82,8 @@ public ConsoleProgressTask(string task, int? totalCount)
             {
                 await Task.Delay(1);
 
+#if RICH_CONSOLE
```
Comment (twsouthwick) [inline src/ApiPort/ConsoleProgressReporter.cs]: Should we create a separate ConsoleProgressREporter? Will this still show up in the platforms that support it?

File: src/ApiPort/ApiPort.project.json (hunk 2 above)
Comment (twsouthwick) [inline src/ApiPort/ApiPort.project.json]: Yup - it was a corner case causing a failure on methods with an arglist parameter. It should be fixed in alpha-17 - are you seeing any tests failing? There was one test failing, so if all tests are passing, it's fixed.

Comment (mjrousos) [inline src/ApiPort/ApiPort.project.json]: Yep, all tests are passing with this one.

File: src/ApiPort/CommandLineOptions.cs (hunk 3 above)
Comment (mjrousos) [inline src/ApiPort/CommandLineOptions.cs]: Sure; I'll switch to that.

File: src/ApiPort/CommandLine/AnalyzeOptions.cs [hunk 5]
```diff
@@ -210,7 +210,21 @@ public FilePathAssemblyFile(string path)
...
+                {
+                    get
+                    {
+                        try
+                        {
+                            return FileVersionInfo.GetVersionInfo(_path).FileVersion;
+                        }
+                        catch (ArgumentException)
+                        {
+                            // Temporary workaround for CoreCLR-on-Linux bug (dotnet/corefx#4727) that prevents get_FileVersion from working on that platform
```
Comment (twsouthwick) [inline src/ApiPort/CommandLine/AnalyzeOptions.cs]: Can you add that the fix should be available with RC2 of .NET Core?

File: src/ApiPort/ConsoleProgressReporter.cs (hunk 4 above)
Comment (mjrousos) [inline src/ApiPort/ConsoleProgressReporter.cs]: I don't think there's any need to make a separate ConsoleProgressReporter since this is already #Ifdef'd out so easily. As it is now, everything works as before on Windows and on Ubuntu/Mac we just don't have the progress reporter's waiting animation (but the tool still seems perfectly usable without it).

**Comment by mjrousos** [comment]
We could put ApiPort.Core in a separate solution and get NuGet packages; I just put it in the existing solution to mirror how our other projects supported .NET Core/Desktop versions. Is there a reason you prefer it to be separated out?

And switching to Autofac seems good. I'll create an issue to take care of that (though not necessarily as part of this PR).

File: src/ApiPort/ConsoleProgressReporter.cs (hunk 4 above)
Comment (twsouthwick) [inline src/ApiPort/ConsoleProgressReporter.cs]: Sounds good

File: src/ApiPort/CommandLine/AnalyzeOptions.cs (hunk 5 above)
Comment (mjrousos) [inline src/ApiPort/CommandLine/AnalyzeOptions.cs]: Yes, good idea.

**Comment by twsouthwick** [comment]
I suggested a separate project since this one introduces so many different configurations. I'm fine with the way it is, though

**Comment by mjrousos** [comment]
Ah, yea, I can see wanting to keep the sln clean. I think it's probably ok as it is, though (I don't look at the sln that often anyhow). I'll leave it for now and we can extract it later, if needed (I did play around with pulling Microsoft.Fx.Portability from NuGet early on and it all works fine, as expected).

File: src/ApiPort/CommandLineOptions.cs (hunk 3 above)
Comment (mjrousos) [inline src/ApiPort/CommandLineOptions.cs]: Actually, I just looked at this one and don't think that FEATURE_ASSEMBLY_LOCATION is right. I remember, now, why I used the general 'NETCORE' define instead of something more specific.

The difference here is whether or not we include the current file's extension in the example command lines in the help messages. I want to drop the extension in the CoreCLR case, since the extension will be .dll and that's not the file that users should be calling. In hosted cases, the user should actually call ApiPort or ApiPort.exe (depending on the platform), so I just leave the extension off.

Comment (twsouthwick) [inline src/ApiPort/CommandLineOptions.cs]: Got it. That makes sense

**Comment by twsouthwick** [comment]
Can you also bump the Microsoft.Fx.Portability.MetadataReader library version since we're increasing the System.Reflection.Metadata reference? Other than that, LGTM

**Comment by twsouthwick** [comment]
Can you update the offline mode section in the docs to mention it is not available on core at the moment?
